walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,58 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    List<TimeRange> optionalEventTimes = new ArrayList<>();
+    List<TimeRange> eventTimes = new ArrayList<>();
+    events.stream()
+          .sorted((e1, e2) -> e1.getWhen().start() - e2.getWhen().start())
+          .forEach(e -> {
+            boolean attendee = !Collections.disjoint(e.getAttendees(), attendees);
+            boolean optional = !Collections.disjoint(e.getAttendees(), optionalAttendees);
+            if (optional || attendee) {
+              optionalEventTimes.add(e.getWhen());
+            } 
+            if (attendee) {
+              eventTimes.add(e.getWhen());
+            }
+          });
+    
+    List<TimeRange> optionalTimes = getOpenTimes(optionalEventTimes, request.getDuration());
+    List<TimeRange> mandatoryTimes = getOpenTimes(eventTimes, request.getDuration());
+    return optionalTimes.size() == 0 && attendees.size() != 0 ? mandatoryTimes : optionalTimes;
+  }
+
+  private List<TimeRange> getOpenTimes(List<TimeRange> eventTimes, long duration) {
+    // Add value to end of list to allow safe "peeking" at the next element in loop.
+    eventTimes.add(TimeRange.fromStartDuration(TimeRange.END_OF_DAY, 0));
+
+    // Jump through day, adding times when we find space.
+    int end = TimeRange.START_OF_DAY;
+    List<TimeRange> meetingTimes = new ArrayList<>();
+    for (int i = 0; i < eventTimes.size() - 1; i++) {
+      meetingTimes.add(TimeRange.fromStartEnd(end, eventTimes.get(i).start(), false));
```

Comment:
A parameter comment would also be useful here:
`meetingTimes.add(TimeRange.fromStartEnd(end, eventTimes.get(i).start(), /* inclusive= */ false));`

Overall, the variable names here impact readability, as the expected format is:`TimeRange.fromStartEnd(start, end, inclusive)`which makes it seem like the first and second parameters are switched. Consider changing the variable names to make it more clear.

Reply:
Good point! I have renamed the variable and added the parameter comment.